src/clusterfuzz/_internal/system/archive.py

```diff
@@ -34,111 +36,316 @@
+
+    Args:
+        member (str): the member name
+        path (str): the path where the member should be extracted.
+        Defaults to None.
+        trusted (bool, optional): whether the archive is trusted. Defaults to
+        False.
+
+    Returns:
+        str: the path to the extracted member
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def open(self, member):
+    """Opens `member`.
+
+    Args:
+        member (str): the member name
+
+    Returns:
+        [IO]: a file-like object that can be `read`.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def close(self) -> None:
+    """Closes the archive.
+    """
+    raise NotImplementedError
+
+  @abc.abstractmethod
+  def extractall(self, path, members=None, trusted=False) -> None:
+    """Extract the whole archive content or the members listed in `members`.
+
+    Args:
+        path (str): the path where the members should be extracted.
+        Defaults to None.
+        members ([str], optional): the member names. Defaults to None.
+        trusted (bool, optional): whether the archive is trusted or not.
+        Defaults to False.
+    """
+    raise NotImplementedError
 
-          yield ArchiveFile(info.filename, info.file_size,
-                            maybe_extract(zip_file.open, info))
+  def try_open(self, member):
+    """Tries to open the archive. Does not throw.
 
-    except (zipfile.BadZipfile, zipfile.LargeZipFile):
-      logs.log_error('Bad zip file %s.' % archive_path)
+    Args:
+        member (str): the member name
 
-  elif archive_type == ArchiveType.TAR:
+    Returns:
+        [IO]: a file-like object that can be `read`.
+    """
     try:
-      if archive_obj:
-        tar_file = tarfile.open(fileobj=archive_obj)
-      else:
-        tar_file = tarfile.open(archive_path)
-
-      for info in tar_file.getmembers():
-        if not file_match_callback(info.name):
-          continue
-
-        yield ArchiveFile(info.name, info.size,
-                          maybe_extract(tar_file.extractfile, info))
-      tar_file.close()
-    except tarfile.TarError:
-      logs.log_error('Bad tar file %s.' % archive_path)
-
-  elif archive_type == ArchiveType.TAR_LZMA:
-    assert archive_obj is None, "LZMAFile doesn't support opening file handles."
+      return self.open(member)
+    except:
+      return None
+
+  def get_first_file_matching(self, search_string):
+    """Gets the name of the first matching member.
+
+    Args:
+        search_string (str): the string to be searched for.
+
+    Returns:
+        str: the member name that matched.
+    """
+    for file in self.list_files():
+      if file.filename.startswith('__MACOSX/'):
+        # Exclude MAC resource forks.
+        continue
+
+      if search_string in file.filename:
+        return file.filename
+    return None
+
+  def extracted_size(self, file_match_callback=None):
+    """Gets the total extracted size of the file matched by
```

Comment:
```suggestion
    """Gets the total extracted size of the files matched by
```

Multiple files can match.